### DISCUSSION TITLE
Best way to make OpenID SSO Server with Pocketbase?

### DISCUSSION THREAD

**Discussion by jahidhsn002** [pocketbase/pocketbase#3740] [Q&A] [answered]
Hi we have quite few micro services. Currently we are running a custom laravel oath2 server as sso. As most of our apis are now already in pocketbase & will be converted soon we want to make a central sso with pocketbase. Most probably some kind of pocketbase as OpenID server with /authorize, /token, /userinfo endpoint. Cannot figure out JWT & BearerToken generation and verification on /userinfo endpoint. There is only user.tokenKey() in the document. Is there a batter way ?

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand what is your current approach of the implementation, but for generating auth tokens you can check the [tokens package](https://pkg.go.dev/github.com/pocketbase/pocketbase/tokens). The only caveat with the JWT generation is that the secret key is constructed from 2 parts - one shared random key stored in the app settings + another random key unique for each auth record (aka. the `TokenKey`). This is done to avoid storing the issued tokens in the database and at the same time to allow invalidating globally all tokens by just changing the settings key part OR if we wanted to invalidate all tokens for a single specific user - we can change only the user's `TokenKey` field.

**With that said, please note that full backward compatibility is not guaranteed until v1 is reached and some of the exposed internals most likely will change in the near future** (_after finalizing the release of my other open source project I'll try to outline a roadmap of things that I intend to change in PocketBase Go APIs to bring it closer to stable v1_).
So if you already have an existing working solution, I don't recommend converting/rewriting it to PocketBase at the moment unless it is just for learning purposes and you don't mind working with poor/missing docs and applying some manual migration steps from time to time.

**Reply by jahidhsn002** [reply]
resolved ...

**Reply by levino** [reply]
Could `pocketbase` not just serve a [`jwks`](https://auth0.com/docs/secure/tokens/json-web-tokens/json-web-key-sets)? I think this is something a bit `auth0` specific, but it is a nice idea and works pretty well. `pocketbase` would just need to host the public keys under `/.well-known/jwks.json` or something. And also potentially not use a custom key signing / salting procedure as described above, but that sounds anyhow a bit unique to me. Usually people just live with the fact that access tokens are valid until expiry which is just not very long...

I would like to use my pocketbase users in another app (nextjs or static hosting on vercel, deny access to certain routes via a middleware for example).

**Reply by ganigeorgiev** [reply]
@levino There are no plans for now to change how the PocketBase auth tokens are issued and validated.

**Reply by levino** [reply]
Understood. My comment was also more of a remark and maybe an idea for the future. I think authentication is a science and also very expensive from a "user acquisition cost" point of view (if users have to create an account conversion usually goes way down). So if my stack consists of multiple back ends (and as far as I can tell, pocketbase has no file hosting, right? So I need to at least have an additional file hoster / cdn) it would be better to be able to use one token for everything.

I do not like the [proposed solution](https://github.com/pocketbase/pocketbase/discussions/1611#discussioncomment-5219912) to use the token for another service as it would put quite some load on the pocketbase instance and kind of reduces the point of a micro service architecture. With `jwks-rsa` the public key is cached on the server, so there are very few requests to the authorization server even if the traffic is very high.

So another solution could be to allow using third party `jwt`s in pocketbase, for example `jwt`s issued by `auth0`. For this one would just have to provide a custom `verifyToken` implementation. Is that possible?

**Reply by ganigeorgiev** [reply]
> and as far as I can tell, pocketbase has no file hosting, right? So I need to at least have an additional file hoster / cdn

PocketBase stores uploaded files either on the local filesystem or in S3 -  https://pocketbase.io/docs/files-handling/.

Custom tokens management can be implemented externally by using the existing [app event hooks](https://pocketbase.io/docs/go-event-hooks/). An auth record is considered authenticated if there is a an auth record in the current request context (aka. when `c.Get("authRecord")` is set; this will slightly change with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) since echo is replaced with a thin wrapper around Go 1.22 std mux).

**Reply by levino** [reply]
FWIW I was more talking about "hosting the react app" than "user generated content". But thanks for the highlight, it is good to know.

**Reply by levino** [reply]
> FWIW I was more talking about "hosting the react app" than "user generated content". But thanks for the highlight, it is good to know.

I missed the `pb_public` part from the [introduction](https://pocketbase.io/docs). There is the static hosting.
